http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/KubernetesServiceDbQueries.scala

```diff
@@ -0,0 +1,288 @@
+      }
+    } yield (cluster, nodepool, app, label, clusterNamespaces, appNamespace, services, ports, disks, diskLabels)
+
+    query.filter {
+      case (_, _, app, _, _, _, _, _, _, _) =>
+        labelQuery
+          .filter(lbl => lbl.resourceId.mapTo[AppId] === app.id && lbl.resourceType === LabelResourceType.app)
+          .filter(lbl =>
+            labelFilter
+              .map { case (k, v) => lbl.key === k && lbl.value === v }
+              .fold[Rep[Boolean]](false)(_ || _)
+          )
+          .length === labelFilter.size
+    }
+  }
+
+  private[db] def aggregateJoinedApp(
+    records: Seq[
+      (KubernetesClusterRecord,
+       NodepoolRecord,
+       AppRecord,
+       Option[LabelRecord],
+       Option[NamespaceRecord],
+       NamespaceRecord,
+       Option[ServiceRecord],
+       Option[PortRecord],
+       Option[PersistentDiskRecord],
+       Option[LabelRecord])
+    ]
+  ): Seq[KubernetesCluster] = {
+
+    val map = records.toList.foldMap {
+      case (clusterRecord,
+            nodepoolRecord,
+            appRecord,
+            labelRecordOpt,
+            clusterNamespaceRecordOpt,
+            appNamespaceRecord,
+            serviceRecord,
+            portRecord,
+            diskRecordOpt,
+            diskLabelOpt) =>
+        val serviceMap = serviceRecord.map(s => Map(s -> Chain.fromSeq(portRecord.toList))).getOrElse(Map.empty)
+        val labelMap = labelRecordOpt.map(labelRecord => labelRecord.key -> Chain(labelRecord.value)).toMap
+        val diskLabelMap = diskLabelOpt.map(labelRecord => labelRecord.key -> Chain(labelRecord.value)).toMap
+        val diskMap = diskRecordOpt.fold[Map[PersistentDiskRecord, Map[String, Chain[String]]]](Map())(disk =>
+          Map(disk -> diskLabelMap)
+        )
+        val appMap = Map(appRecord -> (serviceMap, Chain(appNamespaceRecord), labelMap, diskMap))
+        val nodepoolMap = Map(nodepoolRecord -> appMap)
+
+        Map(clusterRecord -> (nodepoolMap, clusterNamespaceRecordOpt.toList))
+    }
+
+    map.map {
+      case (clusterRec, (nodepoolMap, clusterNamespaces)) =>
+        unmarshalKubernetesCluster(
+          clusterRec,
+          unmarshalNodepoolMap(nodepoolMap),
+          clusterNamespaces.toSet[NamespaceRecord].map(rec => Namespace(rec.id, rec.namespaceName)).toList
+        )
+    }.toSeq
+  }
+
+  private def unmarshalNodepoolMap(
+    nodepools: Map[NodepoolRecord, Map[AppRecord,
+                                       (Map[ServiceRecord, Chain[PortRecord]],
+                                        Chain[NamespaceRecord],
+                                        Map[String, Chain[String]],
+                                        Map[PersistentDiskRecord, Map[String, Chain[String]]])]]
+  ): List[Nodepool] =
+    nodepools
+      .map {
+        case (nodepoolRec, appMap) =>
+          unmarshalNodepool(nodepoolRec, unmarshalAppMap(appMap))
+      }
+      .toSet
+      .toList
```

Comment:
can we get rid of `toSet` here?

Reply:
we need to .toSet this, because the database queries do not GroupBy, and thus duplicates are returned on all tables due to the permutations SQL gives by default on joins